File: 557-reverse-words-in-a-string-iii/557-reverse-words-in-a-string-iii.cpp
class Solution {
public:
    void reverse(string &s, int i, int j){
        while(i <= j){
            swap(s[i++],s[j--]);
        }
    }
    string reverseWords(string s) {
        int n = s.size();
        int i = 0;
        int j = 0;
        while(j <= n){
            if(s[j] == ' ' or j == n){
                reverse(s,i,j-1);
                i = j+1;
                j++;
                continue;
            }
            j++;
        }
        return s;
    }
};

File: 557-reverse-words-in-a-string-iii/557-reverse-words-in-a-string-iii.java
class Solution {
    public String reverseWords(String s) {
        char s1[] = s.toCharArray();
        int i = 0;
        for(int j = 0; j < s1.length; j++){
            if(s1[j] == ' '){
                reverse(s1, i, j-1);
                i = j+1;
            }
        }
        reverse(s1,i, s1.length-1);
        return new String(s1);
    }
    static void reverse(char s1[], int l, int r){
        while(l < r){
            char c = s1[l];
            s1[l] = s1[r];
            s1[r] = c;
            l++;
            r--;
        }
    }
}